MAINT: Translate wright_bessel function to C++

steppi opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message

Depending on your changes, you can skip CI operations and save time and energy: 
http://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skipping

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
https://github.com/scipy/scipy/issues/19404

#### What does this implement/fix?
<!--Please explain your changes.-->
This PR translates the `wright_bessel` function into CUDA compatible C++ along with its dependencies. Following up on previous work like https://github.com/scipy/scipy/pull/19834.

#### Additional information
This branches off of the digamma PR, https://github.com/scipy/scipy/pull/19834, which itself branches off of the `loggamma` PR https://github.com/scipy/scipy/pull/19496. These previous PRs should be merged first before this goes in. I'm just making this now so @izaid can have a chance to review while we wait for the other PRs to get a final look.
<!--Any additional information you think is important.-->

steppi committed:
Get wright_bessel working in CuPy with CUDA

steppi committed:
Change Sinpi/Cospi back to sinpi/cospi

steppi commented:
The failures on 32 bit is because it compares `wright_bessel(a, b, 0)` against `rgamma(b)`, and `wright_bessel` no longer uses the C implementation from `scipy/special/cephes` for this case, but instead the C++ implementation from `scipy/special/special/cephes`. `rgamma` is still using the C implementation, so there's therefore no reason to expect these to be equal any longer.

`mpmath` agrees with the C++ implementation of `rgamma` so I think I'll just swap that one out too.

steppi committed:
Change rgamma ufunc to use cephes C++ translation

h-vetinari reviewed (commented):
A glancing look, nothing more.

steppi commented on scipy/special/special/cephes/zeta.h:
```diff
@@ -0,0 +1,168 @@
+        double a, b, k, s, t, w;
+
+        if (x == 1.0)
+            goto retinf;
```
I'm just copying things over from cephes and making the minimal changes needed to get it to work with Cuda. We can update the control flow later on.

steppi commented on scipy/special/special/cephes/lanczos.h:
```diff
@@ -0,0 +1,112 @@
+
+/* Both lanczos.h and lanczos.c were formed from Boost's lanczos.hpp
+ *
+ * Scipy changes:
+ * - 06-22-2016: Removed all code not related to double precision and
+ *   ported to c for use in Cephes. Note that the order of the
+ *   coefficients is reversed to match the behavior of polevl.
```
Agreed. It's an oversight on my part not to include that in these.

steppi commented on scipy/special/special/cephes/lanczos.h:
```diff
@@ -0,0 +1,112 @@
+/*
+ * Lanczos Coefficients for N=13 G=6.024680040776729583740234375
+ * Max experimental error (with arbitrary precision arithmetic) 1.196214e-17
+ * Generated with compiler: Microsoft Visual C++ version 8.0 on Win32 at Mar 23 2006
```
Yeah. This comment comes from Boost, where this was originally ported from, https://www.boost.org/doc/libs/1_84_0/boost/math/special_functions/lanczos.hpp. For each choice of N and g, you get a different set of coefficient in the expansion, and boost gives several options, noting the accuracy observed in the PhD dissertation *An Analysis of the Lanczos Gamma Approximation* by Glendon Ralph Pugh, https://web.viu.ca/pughg/phdThesis/phdThesis.pdf.

You make a good point that it would make sense to clean up the comments while translating.

rlucas7 replied:
FWIW if you're not familiar with this method-I wasn't until a year or so ago-then you might have no clue where the `G, N` values come from. For this reason I think it's worthwhile to keep the comment here as written, especially because these aren't in the public api and there is no docstring to put the reference into.  Let's leave the comment here as is, for maintainenance and readability sakes.

rlucas7 replied:
If you're not familiar with this method or these documents-I wasn't until a year or two ago-then it is super mysterious where these constants come from. For maintenance time and readability purposes it makes sense to keep these comments.

rlucas7 reviewed (commented):
@steppi two questions:

(1) it looks like `_evalpoly.pxd` is removed, did this already have a corresponding `_evalpoly.h` file in a different migration PR? 

(2) For my own edification where is `SPECFUN_HOST_DEVICE` getting set in the build? 

o/w I can likely approve this one. There are some lingering comments from @h-vetinari on goto statements but we can do a separate PR to remove/cleanup some items. I agree it's easier to isolate those changes in a separate PR-will make any future bisects easier to isolate.

rlucas7 reviewed (commented):
The doc build is failing in gh-actions, do we need to add any docstrings for thjese changes?

steppi commented:
> (1) it looks like `_evalpoly.pxd` is removed, did this already have a corresponding `_evalpoly.h` file in a different migration PR?

Yes, this was removed in the PR translating the `lambertw` function, https://github.com/scipy/scipy/pull/19435/files, but we had to wait until all Cython files using `_evalpoly.pxd` were removed before removing it.

> (2) For my own edification where is `SPECFUN_HOST_DEVICE` getting set in the build?

This gets set here https://github.com/scipy/scipy/blob/62d2af2e13280d29781585aa39a3c5a5dfdfba17/scipy/special/special/config.h#L3-L4, conditionally on whether or not the code is running on CUDA. That's a good file to read to understand how we make it so the headers can be used on CPU and also GPU with CUDA.

> The doc build is failing in gh-actions, do we need to add any docstrings for thjese changes?

I don't think we've touched docs, I thought it was unrelated. I'll rebase on main and push again and see if it still happens.

steppi committed:
EMPTY: [skip actions] [skip cirrus]

steppi committed:
Update header comment

izaid commented on scipy/special/cephes/lanczos.h:
```diff
@@ -130,4 +130,6 @@ static const double lanczos_sum_near_2_d[12] = {
 
+double lanczos_sum_expg_scaled(double x);
+
 #endif
```
You should replace the include guards with #pragma once!

steppi replied:
Ha. Amazed I missed that! Will do.

steppi replied:
Wait. That's just the header in the original cephes library. It's not clear to me why I even touched it. The one in the new C++ library is https://github.com/scipy/scipy/pull/19954/files#diff-2e3af56b6eca119873a12e189a6d166ccfce02406b414a5f099542163cc117b4

steppi replied:
@izaid. Good catch. That was just a file I'd inadvertently changed. I shouldn't have even touched it.

izaid reviewed (approved):
Looks good to me @steppi !

steppi committed:
Remove accidental change to original cephes lanczos.h

steppi committed:
Remove extra blank line

rlucas7 commented on scipy/special/special/wright_bessel.h:
```diff
@@ -0,0 +1,737 @@
+         *
+         * int_0^inf K(a, b, x, r+eps) dr = exp(-eps) int_0^inf exp(-r) Kmod(.., r) dr
+         *
+         * Note the shift r -> r+eps to have integation from 0 to infinity.
```
```suggestion
         * Note the shift r -> r+eps to have integration from 0 to infinity.
```
tiny typo in the comment line

rlucas7 replied:
```suggestion
         * Note the shift r -> r+eps to have integration from 0 to infinity.
```
tiny typo

rlucas7 commented:
This looks ready to merge @steppi I'll leave it open for another day or two if anyone has further comments, plan to merge sometime on Sunday or Monday assuming no further comments.

lucascolley commented:
CI failure looks real

steppi commented:
> CI failure looks real

Interesting. It looks like that job was added only a few weeks ago. I'll have to look into it.

rlucas7 commented:
> > CI failure looks real
> 
> Interesting. It looks like that job was added only a few weeks ago. I'll have to look into it.

[Here](https://github.com/scipy/scipy/actions/runs/8493473991/job/23267584749?pr=19954#step:4:8532) it looks like a [relative tolerance difference of less than `1e-11](https://github.com/scipy/scipy/blob/main/scipy/special/tests/test_data.py#L711)` which we have some test cases of over [here](https://github.com/scipy/scipy/blob/main/scipy/special/tests/test_wright_bessel.py#L76)

It looks like we could put it into that for now in lieue of a precision improvement in the underlying implementation.

steppi commented:
> > > CI failure looks real
> > 
> > 
> > Interesting. It looks like that job was added only a few weeks ago. I'll have to look into it.
> 
> [Here](https://github.com/scipy/scipy/actions/runs/8493473991/job/23267584749?pr=19954#step:4:8532) it looks like a [relative tolerance difference of less than `1e-11](https://github.com/scipy/scipy/blob/main/scipy/special/tests/test_data.py#L711)` which we have some test cases of over [here](https://github.com/scipy/scipy/blob/main/scipy/special/tests/test_wright_bessel.py#L76)
> 
> It looks like we could put it into that for now in lieue of a precision improvement in the underlying implementation.

It's actually different test that's failing, and in some cases it's failing dramatically

```
_____________ test_local[<Data for wright_bessel: wright_bessel>] ______________
[gw1] darwin -- Python 3.11.8 /Users/runner/miniconda3/envs/scipy-dev/bin/python
scipy/special/tests/test_data.py:717: in test_local
    _test_factory(test)
        test       = <Data for wright_bessel: wright_bessel>
scipy/special/tests/test_data.py:725: in _test_factory
    test.check(dtype=dtype)
E   AssertionError: 
E   Max |adiff|: 2.48756e+44
E   Max |rdiff|: 1.15224e+07
E   Bad results (202 out of 3512) for the following points (in output 0):
```

max rdiff on the order of 1e7 is catastrophic. I plan to debug by spinning up an M1 instance on AWS. It also looks like `test_cython_special` is timing out. I think this might relate to changes in the ufunc infrastructure from https://github.com/scipy/scipy/pull/20260. I'll look into that too.

steppi committed:
Fix wright_bessel

steppi committed:
Use sinpi in cephes/gamma.h and make lgam not template function

- lgam uses minimax polynomial calculations which are specific to the
precision used

steppi committed:
Purge specfun's gamma function

steppi commented:
Ok, so there were two test failures. One was `test_cython_special.py` timing out. This actually had nothing to do with `cython_special`, and was actually because `wright_bessel(10.0, 1.0, 10.0)` was getting stuck in an infinite loop. The reason for this was due to mixing unsigned and signed integers in a way that was illegitimate in C++ but seemed to work in Cython. I fixed this by changing all unsigned ints to signed ints in the offending function `special::detail::wb_large_a`.

I was able to reproduce the `scipy.special.tests.test_data` failures on my Linux machine once this was fixed. The reason it was only failing on Mac in CI was because the `cython_special` tests didn't hang on Mac. There's likely some undefined behavior involved which is different between gcc and clang. I was able to reproduce these failures locally after rebasing on `main` and fixing the previous issue. I do not understand, but reverting @izaid's changes to `special::cephes::lgam_sgn` and `special::cephes::lgam` from https://github.com/scipy/scipy/pull/20260 fixed the issue locally.

rlucas7 commented:
This is all green now. Thanks for all the hard work with this one @steppi I'm merging into main now.
